bin/pkg.ml

```diff
@@ -1,90 +1,82 @@
 open Import
 module Lock_dir = Dune_pkg.Lock_dir
+module Fetch = Dune_pkg.Fetch
+module Opam = Dune_pkg.Opam
+module Repo_selection = Opam.Repo_selection
 
-module Lock = struct
-  module Repo_selection = struct
-    module Env = struct
-      module Source = struct
-        type t =
-          | Global
-          | Pure
+let xdg = Xdg.create ~env:Sys.getenv_opt ()
 
-        let to_string = function
-          | Global -> "global"
-          | Pure -> "pure"
+module Opam_repository = struct
+  type t = { url : OpamUrl.t }
 
-        let default = Global
+  let default =
+    { url = OpamUrl.of_string "git+https://github.com/ocaml/opam-repository.git"
+    }
 
-        let term =
-          let all = [ Global; Pure ] in
-          let all_with_strings = List.map all ~f:(fun t -> (to_string t, t)) in
-          let all_strings = List.map all_with_strings ~f:fst in
-          let doc =
-            sprintf
-              "How to initialize the opam environment when taking the opam \
-               repository from a local directory (may only be used along with \
-               the --opam-repository-path option). Possible values are %s. \
-               '%s' will use the environment associated with the current opam \
-               switch. '%s' will use an empty environment. The default is \
-               '%s'."
-              (String.enumerate_and all_strings)
-              (to_string Global) (to_string Pure) (to_string default)
-          in
-          Arg.(
-            value
-            & opt (some (enum all_with_strings)) None
-            & info [ "opam-env" ] ~doc)
-      end
+  let path { url } =
+    let open Fiber.O in
+    let ( / ) = Filename.concat in
+    Fiber.of_thunk (fun () ->
+        let target_dir = Xdg.cache_dir xdg / "dune/opam-repository" in
+        let target = target_dir |> Path.External.of_string |> Path.external_ in
+        let+ res = Fetch.fetch ~checksum:None ~target url in
+        match res with
+        | Ok () -> Ok (Path.to_string target)
```

Comment:
Perhaps we're stringifying the path a bit early?

Reply:
It's a good point, I've converted the function to return a `Path.t` and `Repo_selection` to take a `Path.t`. Oddly enough this seems to work fine, contrary to the large explanatory comment.

@gridbugs, am I missing something?

Reply:
Possibly it's because you're deferring the creation of the path until after the global state it depends on has been initialised.

Reply:
Can you test that it works for relative external paths such as "../opam-repository"?

Reply:
You're right, it doesn't work with relative paths:

```
Description:
  ("Path.External.of_string: relative path given",
  { t = "../../.cache/dune/opam-repository/" })
```

This looks like a deliberate restriction of `Path.External`, but in this case I think relative paths should be supported. So back to `string`?

Reply:
Yes, you can use a string. Or just use `Path.Outside_build_dir.t`.

Note that where I initially pointed out the issue, the stringifying wasn't being done in the command line argument parsing

Reply:
Using `Path.Outside_build_dir.t` I get `Error: path outside the workspace: ../../.cache/dune/opam-repository/ from .` Oddly enough it works fine if I specify an absolute path (that is also outside the workspace), so there seems to be some discrepancy between how it is parsed between relative or absolute paths.

Reply:
Ah yes, I think we just need to normalize such relative paths to `Path.External.t` though.

Reply:
`Path.External.of_filename_relative_to_initial_cwd` seems to be exactly what I am looking for.

Reply:
Can you refer to paths inside the current project with that function? It's not a common use case but comes in very handy for writing tests.

Reply:
Yep, it seems to work fine. I was able to adjust all the test cases just fine and they work well.